@keeramis Looks like this is failing on Gen 4 platforms:
```
make -s clean all PLATFORM_ID=32 COMPILE_LTO=n DEBUG_BUILD=n MODULAR=y USE_SWD_JTAG=n USE_SWD=n APP=tinker
  /firmware/modules/tron/system-part1/makefile /firmware/modules/tron/user-part/makefile
./src/rtl872x/usbd_device.cpp: In function 'char* {anonymous}::device_state_as_string(char*)':
./src/rtl872x/usbd_device.cpp:53:16: error: 'security_mode_get' was not declared in this scope
   53 |     int mode = security_mode_get(nullptr);
      |                ^~~~~~~~~~~~~~~~~
./src/rtl872x/usbd_device.cpp:56:19: error: 'security_mode_is_overridden' was not declared in this scope
   56 |       memcpy(buf, security_mode_is_overridden() ? "sm=s" : "sm=o", 4);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/bootloader/src/nRF52840/usbd_device.cpp
+++ b/bootloader/src/nRF52840/usbd_device.cpp
@@ -187,6 +187,7 @@ const uint8_t USBD_MsftStrDesc[] = {
   )
 };
 
+
 char* device_id_as_string(char* buf) {
   uint8_t deviceId[HAL_DEVICE_ID_SIZE] = {};
   unsigned deviceIdLen = hal_get_device_id(deviceId, sizeof(deviceId));
@@ -196,15 +197,13 @@ char* device_id_as_string(char* buf) {
 
 char* device_state_as_string(char* buf) {
     int mode = security_mode_get(nullptr);
-    const char* state;
     
     if (mode == MODULE_INFO_SECURITY_MODE_NONE) {
-        state = security_mode_is_overridden() ? "sm=s" : "sm=o";
+      memcpy(buf, security_mode_is_overridden() ? "sm=s" : "sm=o", 4);
     } else {
-        state = "sm=p";
+      memcpy(buf, "sm=p", 4);
     }
-    
-    memcpy(buf, state, 4); // "sm=?" is always 4 bytes
+
     return buf;
 }
 
@@ -770,7 +769,7 @@ const uint8_t* NrfDevice::getString(SetupRequest* r, uint16_t* len) {
     case STRING_IDX_DEVICE_STATE: {
       /* No conversion to UTF-16 */
       char deviceStateStr[8] = {0};
-      return getUnicodeString(device_state_as_string(deviceStateStr), sizeof(deviceStateStr)-1, len, false);
+      return getUnicodeString(device_state_as_string(deviceStateStr), sizeof(deviceStateStr), len, false);
     }
     default: {
       if (drv_) {

--- a/hal/src/rtl872x/usbd_device.cpp
+++ b/hal/src/rtl872x/usbd_device.cpp
@@ -26,6 +26,7 @@
 #include <algorithm>
 #include "appender.h"
 #include "device_code.h"
+#include "security_mode.h"
 
 using namespace particle::usbd;
 
@@ -51,15 +52,13 @@ char* device_id_as_string(char* buf) {
 
 char* device_state_as_string(char* buf) {
     int mode = security_mode_get(nullptr);
-    const char* state;
     
     if (mode == MODULE_INFO_SECURITY_MODE_NONE) {
-        state = security_mode_is_overridden() ? "sm=s" : "sm=o";
+      memcpy(buf, security_mode_is_overridden() ? "sm=s" : "sm=o", 4);
     } else {
-        state = "sm=p";
+      memcpy(buf, "sm=p", 4);
     }
-    
-    memcpy(buf, state, 4); // "sm=?" is always 4 bytes
+
     return buf;
 }
 
@@ -375,7 +374,7 @@ int Device::getString(unsigned id, uint16_t langId, uint8_t* buf, size_t len) {
     case STRING_IDX_DEVICE_STATE: {
       /* No conversion to UTF-16 */
       char deviceStateStr[8] = {0};
-      return getUnicodeString(device_state_as_string(deviceStateStr), sizeof(deviceStateStr)-1, len, false);
+      return getUnicodeString(device_state_as_string(deviceStateStr), sizeof(deviceStateStr), buf, len);
     }
     default: {
         for (auto& cls: classDrivers_) {

